docs/proposals/cross-mesh-ux.md

```diff
@@ -0,0 +1,229 @@
+This may be a more generally useful abstraction given potential
+other kinds of `sources`. For example, one for filtering based on JWT token:
+
+```
+  - routeName: cross-mesh
+    action: allow
+    sources:
+    - user:
+        email: abc@xyz.com
+```
+
+#### Inline `MeshGatewayRoute`
+
+Another option is to add cross-mesh access control to the `MeshGatewayRoute`
+resource, as another type of `conf.http.rules.matches` entry or
+potentially at the top level under `conf.http`.
+
+```
+type: MeshGatewayRoute
+mesh: expose
+name: edge-gateway-route
+selectors:
+  - match:
+      kuma.io/service: edge-gateway
+      name: mesh-listener
+conf:
+  http:
+    mesh:                    # new entry type
+      - name: consume
+        tags:
+          kuma.io/service: client
+    rules:
+      - matches:
+          # ALL match types must succeed
+          - path:
+              match: PREFIX
+              value: /api
+          - mesh:                    # and/or a new entry type here
+            - name: consume
+              tags:
+                kuma.io/service: client
+        backends:
+          - destination:
+              kuma.io/service: server
+      - matches:    # ALL matches must succeed
+          - path:
+              match: PREFIX
+              value: /api/v2
+          - mesh:
+            - name: *
+        backends:
+          - destination:
+              kuma.io/service: server
+```
+
+The advantage of this option is probably reduced complexity for less complicated
+cases.
+
+### `consume` UX
+
+In order for `consume` services to communicate with `expose`, we need to give
+those services a way to contact the `MeshGateway`. A `MeshGateway` doesn't have
+an address or concrete identity beyond the identities of the `kuma-dp` proxies that
+serve the gateway.
+
+#### Virtual outbounds
+
+We can add support for `MeshGateway` listeners to `VirtualOutbounds` to give applications
+a way to address a `MeshGateway` listener from another mesh.
+
+The `VirtualOutbound` selects `Dataplane`s and uses them to generate outbounds.
+These generated outbounds are added to _all_ `Dataplane` resources.
+
+```
+---
+type: VirtualOutbound
+mesh: consume
+name: test
+selectors: # this VirtualOutbound will be built using the matched data plane proxies and will be added for all data plane proxies
+  - match:
+      kuma.io/service: edge-gateway
+      name: mesh-listener
+conf:
+  host: "gateway.{{.gateway.mesh}}.mesh"
+    # or host: "{{.gateway.listener.hostname}}"
+  port: "{{.gateway.listener.port}}"
+  parameters:
+    - name: service
+      tagKey: "kuma.io/service"
+    - name: gateway
+      gatewayListener: {}
+```
+
+At the moment, the `selectors` are only run for `Dataplane` resources whereas we
+need support for `crossMesh`-enabled `MeshGateway` listeners (of all `Mesh`es) as well. In the
+implementation, this means keeping track of all `MeshGateway`-serving `Dataplane`s
+of _any mesh_ when generating a `VirtualOutbound` of _any mesh_.
```

Comment:
This feels really hacky... I wish there would be a better way to handle this.

I'm thinking maybe adding a type to a vob and we'd have:

- subset
- cross-mesh
- external-service (in the future I guess, or never :))

subset would be the default and what we do atm.

remote-mesh would be a selector on a remote mesh service (restricted to gateway in the first version).
The question there is how do you pass information about changing services from another mesh and not end up with any mesh change triggering a refresh of all other bound meshes. Maybe we need a configmap or something like this to avoid this?

Reply:
So the idea being the same just making the fact that it's a gateway listener match explicit?

> The question there is how do you pass information about changing services from another mesh and not end up with any mesh change triggering a refresh of all other bound meshes.

Won't it only trigger a refresh if a MeshGateway changes?

Reply:
Sorry I meant changes the gateway instances need to trigger changes of all dps accessing the gateway in other meshes. As this is the case how do you watch for these changes without reconciling all the meshes too often?

Reply:
Yes, I wonder whether this is a problem in practice. I would suppose the amount of MeshGateway Dataplanes and changing to the `MeshGateway` is constant/grows slowly depending on the number of services in that Mesh, so maybe there won't be that many changes? Maybe we could require the `consume` Mesh to explicitly _opt-in_ to using other MeshGateways which would at least make this a conscious choice.

Reply:
Yes I'm more concerned in envs with a lot of meshes where the change to any gateway will trigger refreshing a lot of things.
I think it's fine to start without and just to make a note in the proposal that this is a place we may want to improve

Reply:
Sounds good:
* using automatic `.hostname` virtual outbound
* icluded a note about performance